wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -245,6 +246,74 @@ public void stop(String workspaceId) throws NotFoundException, ServerException,
     }
 
     /**
+     * Adds machine into running workspace.
+     * This method do not touch workspace configuration.
+     * Just adds machine to workspace runtime and destroy it on workspace stop.
+     *
+     * @param machine
+     *         machine to add to specified runtime
+     * @throws NotFoundException
+     *         when workspace with specified id not running or not exists
+     * @throws ServerException
+     *         when application server is stopping
+     * @throws ConflictException
+     *         when workspace is not running
+     */
+    public void addMachine(MachineImpl machine) throws NotFoundException, ServerException, ConflictException {
+        ensurePreDestroyIsNotExecuted();
+
+        String workspaceId = machine.getWorkspaceId();
+
+        rwLock.writeLock().lock();
+        try {
+            ensurePreDestroyIsNotExecuted();
+            final RuntimeDescriptor descriptor = descriptors.get(workspaceId);
+            if (descriptor == null) {
+                throw new NotFoundException("Workspace with id '" + workspaceId + "' is not running.");
+            }
+            if (descriptor.getRuntimeStatus() != WorkspaceStatus.RUNNING) {
+                throw new ConflictException("Cannot add machine " + machine.getId() + " to not running workspace.");
+            }
+
+            descriptor.getRuntime().getMachines().add(machine);
+        } finally {
+            rwLock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Removes machine from running workspace.
+     * This method do not touch workspace configuration and is opposite to {@link #addMachine(MachineImpl)}
+     * Does nothing if remove nonexistent machine
+     *
+     * @param machine
+     *         machine to remove from specified runtime
+     * @throws NotFoundException
+     *         when workspace with specified id not running or not exists
+     */
+    public void removeMachine(MachineImpl machine) throws NotFoundException {
+        String workspaceId = machine.getWorkspaceId();
+
+        rwLock.writeLock().lock();
+        try {
+            final RuntimeDescriptor descriptor = descriptors.get(workspaceId);
+            if (descriptor == null) {
+                throw new NotFoundException("Workspace with id '" + workspaceId + "' is not running.");
+            }
+
+            List<MachineImpl> machines = descriptor.getRuntime().getMachines();
+            for (int i = 0; i < machines.size(); i++) {
+                if (machine.getId().equals(machines.get(i).getId())) {
+                    machines.remove(i);
```

Comment:
Hi,

Indeed it is quite clear that in this particular case it is not a good idea to walk through the List using index. 

However the explanation looks to me a bit confusing so I would like to clarify a few things. The main argument here is a worst-case complexity of `get` method for different implementations of `List`, using asymptotic notation for `LinkedList` we have _O(n)_ while for `ArrayList` we have _O(1)_. This gives us total estimations: _O(n*n)_ for `LinkedList` and _O(n)_ for `ArrayList` for our code snippet. At the same time using iterator you will have _O(n)_ for both implementations. 

P.S. Still can't figure out where did you get such formula: _1/2 * N * (1 + N)_. Does that correspond to some average-case complexity measurement?